### PR TITLE
Interaction resumption

### LINKED ISSUE #5138: Add multi-selection support to dropdown widget
The WidgetMultiSelection should work almost the same way as WidgetSelection, but (obviously) allow selecting multiple choices (it does not close after click at choice), and the picked items are put in List - and this list should be displayed with vector widget, because we want to allow reordering picked choices.

<hr />

This task is automatically imported from the old Task Issue Board and it was originally created by Paweł Grabarz.
Original issue is [here](https://www.pivotaltracker.com/story/show/184012753).

> Step 3: Support for multi-selection
> 
> Delivery requirements:
> 
> The implementation should be refined to allow a multi-selection of entries. The information if the widget should allow multi-selection should be delivered from the engine together with the information that a particular input should be displayed as a drop-down widget.
> 
> See the widget metadata design to learn more: https://github.com/enso-org/design/blob/main/epics/widgets/drop-downs/dynamic-widgets.md

### PR DESCRIPTION
**Pull request by kazcw**
### Pull Request Description

Prevent interactions such as an open dropdown from being disrupted by the temporary state of absent type information occurring after an edit.

https://github.com/enso-org/enso/assets/1047859/bcc7fa02-847f-4cd0-b9fc-3186a97c5816

All widget-edit interactions that are active when a component is to be unmounted save state and are suspended. When a new component defining a `WidgetEditHandler` is instantiated, if the component is found to be equivalent to a component that was suspended, and no other interaction has been initiated in the interim, the interaction is restarted using the suspended state.

### Important Notes

- Fix a bug caused by a variable tracking an interaction's active state getting out of sync with the interaction. `WidgetEditHandler` now provides a reactive `active` property; using it is simpler and avoids this type of bug in the future.
- Fix a flickering bug that sometimes made it hard to open dropdowns by clicking the arrow.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [ ] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [ ] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

File: app/gui2/src/components/GraphEditor/widgets/WidgetSelection.vue [hunk 1]
```diff
@@ -99,12 +98,14 @@ class ActionTag {
 type ExpressionFilter = (tag: ExpressionTag) => boolean
 function makeExpressionFilter(pattern: Ast.Ast | string): ExpressionFilter | undefined {
   const editedAst = typeof pattern === 'string' ? Ast.parse(pattern) : pattern
+  const editedCode = pattern instanceof Ast.Ast ? pattern.code() : pattern
   if (editedAst instanceof Ast.TextLiteral) {
+    if (props.input.value instanceof Ast.TextLiteral && props.input.value.code() === editedCode)
+      return undefined
```
Comment (farmaazon) [inline app/gui2/src/components/GraphEditor/widgets/WidgetSelection.vue]: AFAIK the design was, the filtering is always applied after any user's change, even if these changes come back to initial value. And why this is only in text literal branch? 

And why do we need to check if props.input.value is TextLiteral? If the code matches, it has to be literal, doesn't it?

File: app/gui2/src/providers/widgetRegistry/editHandler.ts [hunk 2]
```diff
@@ -46,102 +198,117 @@ export interface WidgetEditInteraction extends Interaction {
...
-    private parent?: WidgetEditHandler,
-    private interactionHandler = injectInteractionHandler(),
-    private widgetTree: { currentEdit: WidgetEditHandler | undefined } = injectWidgetTree(),
+    widgetId: WidgetId,
+    hooks: WidgetEditHooks,
+    parent: WidgetEditHandler | undefined,
+    portEditor: PortEditor,
+    widgetTree: {
+      currentEdit: WidgetEditHandler | undefined
+    } = injectWidgetTree(),
```
Comment (farmaazon) [inline app/gui2/src/providers/widgetRegistry/editHandler.ts]: I saw that we use `private readonly widgetId: WidgetId` syntax in constructor parameters elsewhere, so it's both declared and set as property.

File: app/gui2/src/util/callTree.ts [hunk 3]
```diff
@@ -13,89 +13,123 @@ export const enum ApplicationKind {
...
     }
   }
+}
+
+/**
+ * Information about an argument that doesn't have an assigned value yet, therefore are not
+ * represented in the AST.
+ */
+export class ArgumentPlaceholder extends Argument {
+  public declare callId: string
```
Comment (farmaazon) [inline app/gui2/src/util/callTree.ts]: Why this `declares` is needed? I thought it's obvious it's string from the parent class?

File: app/gui2/src/providers/widgetRegistry/editHandler.ts [hunk 4]
```diff
@@ -46,102 +198,117 @@ export interface WidgetEditInteraction extends Interaction {
  * is bound to two children, and one of them starts editing while the other is edited, the parent
  * will receive `cancel` feedback from the latter and then `start` from the former.
  *
- * **The `click` handler is a special case:** it will be called only on top-most parent, but its
- * handler may decide to delegate it further by calling child's handler passed as an additional
- * argument
+ * **The `click` handler is a special case:** it acts as a capture-mode event handler; it is called on
+ * the top-most widget, and a widget may choose to delegate to its child (if any) by returning false.
```
Comment (farmaazon) [inline app/gui2/src/providers/widgetRegistry/editHandler.ts]: `click` was renamed to `pointerdown` in this PR.

File: app/gui2/src/providers/widgetRegistry/editHandler.ts [hunk 5]
```diff
@@ -46,102 +198,117 @@ export interface WidgetEditInteraction extends Interaction {
...
         noLongerActive()
-        innerInteraction.end?.(portId)
-        parent?.interaction.end?.(portId)
+        hooks.end?.(origin)
       },
-      addItem: () => {
-        return (innerInteraction.addItem?.() || parent?.interaction.addItem?.()) ?? false
+      pointerdown: (event, navigator) => {
+        if (!hooks.pointerdown) return false
+        return hooks.pointerdown?.(event, navigator)
```
Comment (farmaazon) [inline app/gui2/src/providers/widgetRegistry/editHandler.ts:75]: Where the propagation is done?

File: app/gui2/src/providers/widgetRegistry/editHandler.ts [hunk 6]
```diff
@@ -1,24 +1,176 @@
...
+  cancel() {}
+}
+
+class PortEditInteraction implements Interaction {
+  readonly portId: PortId
+  readonly argId: string | undefined
+  readonly active: Ref<boolean>
+  private readonly interactionHandler
+  private readonly resumable: PortEditResumeData
+  private readonly interactions = new Array<PortEditSubinteraction>()
```
Comment (farmaazon) [inline app/gui2/src/providers/widgetRegistry/editHandler.ts]: What is a difference between `PortEditInteraction` and `PortEditSubinteraction`?

File: app/gui2/src/providers/widgetRegistry/editHandler.ts [hunk 7]
```diff
@@ -1,24 +1,176 @@
...
+
+  getResumeHook(widgetId: WidgetId) {
+    return this.resumable.has(widgetId) ? this.resumable.get(widgetId) ?? (() => {}) : undefined
+  }
+}
+
+interface Endable {
+  end?(origin: WidgetId): void
+}
+/** A child interaction of a @{link PortEditInteraction} */
```
Comment (farmaazon) [inline app/gui2/src/providers/widgetRegistry/editHandler.ts]: What is "child interaction"? Is it the same as "child's interaction"? In this page docs, the only parent/child relation is this of Widgets.

File: app/gui2/src/providers/widgetRegistry/editHandler.ts [hunk 8]
```diff
@@ -1,24 +1,176 @@
...
+   *  implemented, the returned `resume` function will be called after starting an interaction for this reason. It may
+   *  use information captured at suspension time; however note that `resume` will be called in a different component
+   *  instance from `suspend`.
+   */
+  suspend?(): { resume: () => void }
+}
+
+/** @internal Public for unit testing.
+ *  Obtains a top-level interaction to edit a port (see {@link PortEditInteraction}), which may be a pre-existing
+ *  ongoing interaction, or a newly-started interaction. */
```
Comment (farmaazon) [inline app/gui2/src/providers/widgetRegistry/editHandler.ts]: I think the name does not quite reflect the usage.

**Review by farmaazon — COMMENTED** [review]
I'd like to see a bit more docs for editHandler, as I don't quite grasp what is the relation between all those classes

File: app/gui2/src/components/GraphEditor/widgets/WidgetSelection.vue (hunk 1 above)
Comment (kazcw) [reply app/gui2/src/components/GraphEditor/widgets/WidgetSelection.vue]: Removed.

**Comment by kazcw** [comment]
I rearranged the file to introduce concepts in a more logical order, and added docs, especially to `WidgetEditInteraction`, which is the most important class besides `WidgetEditHandler` and its public interface.

File: app/gui2/src/providers/widgetRegistry/editHandler.ts [hunk 9]
```diff
@@ -198,27 +37,20 @@ export class PortEditor {
...
-    currentEdit: WidgetEditHandler | undefined
-  }
 
   constructor(
-    widgetId: WidgetId,
-    hooks: WidgetEditHooks,
-    parent: WidgetEditHandler | undefined,
-    portEditor: PortEditor,
-    widgetTree: {
+    private readonly widgetId: WidgetId,
```
Comment (farmaazon) [inline app/gui2/src/providers/widgetRegistry/editHandler.ts:49]: Having this, the initializations (lines 57 onwards) are redundant.

**Review by farmaazon — APPROVED** [review]

**Pull request merged**